filters/include/pcl/filters/experimental/cartesian_filter.h

```diff
@@ -0,0 +1,212 @@
+  const std::size_t ijk2 = std::floor(pt.z * inverse_leaf_size[2]) - min_b[2];
+  return ijk0 + ijk1 * divb_mul_y + ijk2 * divb_mul_z;
+}
+
+/** \brief Check if the leaf size is too small, given the range of the point cloud
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] inverse_leaf_size 1/leaf_size of dimension x
+ * \param[in] inverse_leaf_size 1/leaf_size of dimension y
+ * \note Return false if hash range is larger than the range of size_t, as the
+ * unsigned int "wrap around" effect will occur during hashing a point.
+ */
+std::size_t
+checkHashRange(const Eigen::Vector4f& min_p,
+               const Eigen::Vector4f& max_p,
+               const float inverse_leaf_size_x,
+               const float inverse_leaf_size_y)
+{
+  const float max_float = std::numeric_limits<float>::max();
+  // Overflow when
+  // a > 0 && b < 0 && a - b > max
+  // -> - b > max - a
+  if ((max_p[0] > 0. && min_p[0] < 0. && -min_p[0] > max_float - max_p[0]) ||
+      (max_p[1] > 0. && min_p[1] < 0. && -min_p[1] > max_float - max_p[1]))
+    return 0;
+
+  const float range_x = max_p[0] - min_p[0];
+  const float range_y = max_p[1] - min_p[1];
+
+  const std::size_t max_size = std::numeric_limits<std::size_t>::max();
+  // Check the number of bins of each dimension
+  if (range_x > (max_size - 1) / inverse_leaf_size_x ||
+      range_y > (max_size - 1) / inverse_leaf_size_y)
+    return 0;
+
+  const std::size_t dx =
+      static_cast<std::size_t>(std::floor(range_x * inverse_leaf_size_x)) + 1;
+  const std::size_t dy =
+      static_cast<std::size_t>(std::floor(range_y * inverse_leaf_size_y)) + 1;
+
+  // Check hashing range
+  if (dy > max_size / dx)
+    return 0;
+  else
+    return dx * dy;
+}
+
+/** \brief Check if the leaf size is too small, given the range of the point cloud
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] min_b the minimum bin coordinate
+ * \param[in] inverse_leaf_size 1/leaf_size of dimension x
+ * \param[in] inverse_leaf_size 1/leaf_size of dimension y
+ * \param[in] inverse_leaf_size 1/leaf_size of dimension z
+ * \note Return false if hash range is larger than the range of size_t, as the
+ * unsigned int "wrap around" effect will occur during hashing a point.
+ */
+std::size_t
+checkHashRange(const Eigen::Vector4f& min_p,
+               const Eigen::Vector4f& max_p,
+               const float inverse_leaf_size_x,
+               const float inverse_leaf_size_y,
+               const float inverse_leaf_size_z)
+{
+  const float max_float = std::numeric_limits<float>::max();
+  // Overflow when
+  // a > 0 && b < 0 && a - b > max
+  // -> - b > max - a
+  if ((max_p[0] > 0. && min_p[0] < 0. && -min_p[0] > max_float - max_p[0]) ||
+      (max_p[1] > 0. && min_p[1] < 0. && -min_p[1] > max_float - max_p[1]) ||
+      (max_p[2] > 0. && min_p[2] < 0. && -min_p[2] > max_float - max_p[2]))
```

Comment:
Would Eigen help here (and elsewhere in this function)?

Reply:
Done, most of the calculations are implemented with eigen operations now, except for `dx` `dy` `dz` (eigen don't have unsigned int type)